### ISSUE TITLE
High concurrency throws runtime error: bad request

### ISSUE THREAD

**Issue by lasrado** [openresty/lua-resty-redis#165]
I am aware of the issues #44, #104 etc. Though this might be related, the issue is not exactly the same.

With the resty library, the redis connections work fine until the transactions are upto 30TPS. But once the number of transactions increase, we notice that around 1% requests fail with the below error.

lua coroutine: runtime error: /usr/local/openresty/lualib/resty/redis.lua:253: bad request

stack traceback:
coroutine 0:
[C]: in function 'send'
/usr/local/openresty/lualib/resty/redis.lua:253: in function 'get'

Is there any limit on number of concurrent requests for the library? If you have a better way to handle this, please let me know.

**Comment by im-mdodd** [comment]
did you ever get to the bottom of this. I see to be having the same issue?

**Comment by agentzh** [comment]
It's a common usage mistake in your Lua code. Never ever share any cosocket objects (or any objects wrapping cosockets, like `resty.redis` objects) across the request handler boundary! Make sure you have dedicated cosocket objects for each individual nginx request. This `bad request` error is a indicator for such cosocket mis-sharing and misuse.

**Comment by rushi47** [comment]
@agentzh  Hope you are doing great :),
I followed your comments and the issues you have solved for #44 and #104.
But am still getting the same issue, for sure its because of my poor understanding and apologise for the same in advance.
But am struggling to understand "resty.redis"  part in the comment which you have mentioned in all the issues.

Below is the snippet of My Code which i tried and am still facing the issue.

**Problem** : _I want to calculate the time redis takes to connect and do operation if its more than
0.100 i log it. To do that, i have overridden every method (Sharing only two here - new & get | And i Know there must be some better approach  than this would love to know that too )._
 

**- redis_extended.lua**

```
local redis_import = require "resty.redis"
local red_mod = {}
function red_mod:new(o, ngx)
	o = o or {}
	setmetatable(o, self) -- http://lua-users.org/wiki/MetamethodsTutorial
	self.__index = self
	self.ngx = ngx
	self.redis = redis_import:new()
	return o
end

function red_mod:get(key)
	local start_time = self.ngx.now()

 --  local redis = require "resty.redis" 

> if i uncomment above line and use below redis:get instead of self.get 
>   everything works fine 
> (this is what i think from initial steps and am getting more confused
>  because of this cause at this point am not even initialising the object)

	local ok, err = self.redis:get(key)
	local end_time = self.ngx.now() - start_time
	if end_time > 0.100 then
		self.ngx.log(self.ngx.ERR, "!@@! Execution Time for REDIS:GET Operation :- ", end_time)
	end
	return ok, err
end

return red_mod
```

And am calling this in the other modules, which is part of nginx co-routine. So if above code is used in one module it doesnt break as soon as i use this in other modules it fails.
```
       local redis = require "eg-modules.redis_extended"
        local red = redis:new(nil, ngx)
	red:set_timeout(1000) -- 1 sec
	local ok, err = red:connect(redis_health_check_host, redis_health_check_port)
	if not ok then
		ngx.log(ngx.ERR,"failed to connect: ", err)
		return
	end
	local key = module.get_request_key(http_x_egnyte_request_id)
	**-- Thing breaks here, in Get call**
        local ok, err = red:get(key .. '.bck')
```

Everything works until the red:get call arrives, and at this point it throws the error.
It will be great help, if you can point out in the code where i am going  wrong.
And on how to achieve this using any code snippet :-

`Never ever share any cosocket objects (or any objects wrapping cosockets, like resty.redis objects) across the request handler boundary! `

----
Out of Context-
And to have understanding of nginx in depth & lua (so that i dont ask such dumb questions) can you share in links (books, videos, docs).
Thanks a lot in advance :) .

**Comment by agentzh** [comment]
@rushi47 Please do not cross-post. It's annoying. Thanks for your cooperation.
